### ISSUE TITLE
Creating an AbstractImage from raw png bytes

### ISSUE THREAD

**Issue by habemus-papadum** [timholy/Images.jl#279]
Hi -- Is there a way to create an Image from png data that is stored in memory (without dumping the data to a temp file)?  Thanks!

**Comment by timholy** [comment]
Yes, see https://github.com/timholy/Images.jl/blob/master/doc/function_reference.md#image-construction.

**Comment by lilinjn** [comment]
My apologies, I'm still not getting it -- I've looked at the docs as well as the code...

To clarify, my in memory data is in PNG format in an Uint8 array and I am looking to convert it to a matrix of colors i.e. an Image.  

As far as I can tell, imread is the function I should be using, and there are methods that take filenames (which I'm hoping to avoid) and streams which would be great as I could use IOBuffer, except the stream version needs a filetype and there doesn't seem to be an ImageFileType for PNG (the only supported file types for reading via stream seem to be Imagine and SIF)

A self contained example of what I am hoping might be possible:

````
g = "digraph foo {1 -> 2;3 -> 2;3 -> 1}"
o,i,p = readandwrite(`dot -Tpng`)  ## use grpahviz to create a png
write(i,g)                                        ## write the input
close(i)
image = imread(o,??What Goes Here??)
````


For reference (but has nothing to do with Images)  the following works perfectly fine in IJulia

````
g = "digraph foo {1 -> 2;3 -> 2;3 -> 1}"
o,i,p = readandwrite(`dot -Tpng`)
write(i,g)
close(i)

d = readbytes(o)
display("image/png",d)
````

(This is quick and dirty example just for clarification)

Thanks (fyi, I used Images to create my github avatar!)

**Comment by timholy** [comment]
Now I get it---sorry.

Probably the easiest would be `img = imread(o, Images.ImageMagick)`. If you're on OSX you could alternatively use the OSXnative reader.

BTW, nice avatar!

**Comment by lilinjn** [comment]
Hi -- Unfortunately, that method does not exist:
`image = imread(o,Images.ImageMagick)`
````
`imread` has no method matching imread(::Pipe, ::Type{ImageMagick})
while loading In[57], in expression starting on line 5
````

And similarly for OSXNative.  

Tracing through the implementation, the ImageMagick version is based on reading from files on disk in a non-trivial way and looks like it would take some effort to make it support streams (honestly, doing the ugly thing of just dumping the stream to a temp file to be deleted later is probably the simplest path for handling the Imagemagick case...)  For the OSXNative, the situation is much simpler -- basically refactoring the code so that one could create a CGImageSource not only from url, but also raw bytes e.g. a CGData.  I will send in pull requests and you can have a look.  Thanks!

**Comment by timholy** [comment]
Presumably you could add a wrapper for ReadImageBlob? http://www.imagemagick.org/api/MagickWand/magick-image_8c.html

**Comment by lilinjn** [comment]
Of course -- Thanks for pointing that out.

**Comment by timholy** [comment]
Fixed by #365
